File: backend/app/schemas/demo_experience.py
"""Public contracts for guided onboarding, live demo, and reporting."""

from __future__ import annotations

from datetime import datetime
from enum import StrEnum
from typing import Annotated
from uuid import UUID

from pydantic import (
    BaseModel,
    ConfigDict,
    EmailStr,
    Field,
    StringConstraints,
    model_validator,
)

SafeKey = Annotated[
    str, StringConstraints(min_length=8, max_length=128, pattern=r"^[A-Za-z0-9._:-]+$")
]


class ImportShape(StrEnum):
    LONG = "long"
    WIDE = "wide"


class ImportMapping(BaseModel):
    model_config = ConfigDict(frozen=True)

    shape: ImportShape
    timestamp: str = Field(min_length=1, max_length=255)
    machine: str = Field(min_length=1, max_length=255)
    sensor: str | None = Field(default=None, min_length=1, max_length=255)
    value: str | None = Field(default=None, min_length=1, max_length=255)
    features: list[str] = Field(default_factory=list, max_length=128)
    target: str | None = Field(default=None, min_length=1, max_length=255)
    operating_state: str | None = Field(default=None, min_length=1, max_length=255)
    batch_or_shift: str | None = Field(default=None, min_length=1, max_length=255)

    @model_validator(mode="after")
    def validate_shape(self) -> ImportMapping:
        if self.shape == ImportShape.LONG and (not self.sensor or not self.value):
            raise ValueError("Long format requires sensor and value columns.")
        if self.shape == ImportShape.WIDE and not self.features:
            raise ValueError("Wide format requires at least one feature column.")
        columns = [
            self.timestamp,
            self.machine,
            self.sensor,
            self.value,
            self.target,
            self.operating_state,
            self.batch_or_shift,
            *self.features,
        ]
        values = [item.casefold() for item in columns if item]
        if len(values) != len(set(values)):
            raise ValueError("A source column cannot be mapped more than once.")
        return self


class QualityIssue(BaseModel):
    model_config = ConfigDict(frozen=True)

    code: str
    severity: str
    count: int = Field(ge=0)
    meaning: str
    correction: str
    samples: list[int] = Field(default_factory=list, max_length=20)


class DataImportResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    company_id: UUID
    filename: str
    media_type: str
    size_bytes: int
    status: str
    delimiter: str
    has_header: bool
    mapping: dict[str, object]
    preview: list[dict[str, object]]
    quality_report: dict[str, object]
    error_samples: list[dict[str, object]]
    total_rows: int
    imported_rows: int
    rejected_rows: int
    progress_percent: int
    warnings_accepted: bool
    created_at: datetime
    updated_at: datetime
    completed_at: datetime | None


class ImportListResponse(BaseModel):
    items: list[DataImportResponse]
    total: int
    limit: int
    offset: int


class ImportMappingRequest(BaseModel):
    mapping: ImportMapping
    delimiter: str = Field(default=",", min_length=1, max_length=1)
    has_header: bool = True


class ConfirmImportRequest(BaseModel):
    accept_warnings: bool = False


class GuidedUseCase(StrEnum):
    PREDICTIVE_MAINTENANCE = "predictive_maintenance"
    ENERGY_MONITORING = "energy_monitoring"
    QUALITY_PREDICTION = "quality_prediction"
    CONDITION_MONITORING = "condition_monitoring"


class TrainingProfile(StrEnum):
    FAST_DEMO = "fast_demo"
    BALANCED = "balanced"
    THOROUGH = "thorough"


class GuidedTemplateResponse(BaseModel):
    use_case: GuidedUseCase
    label: str
    task_type: str
    supported_profiles: list[TrainingProfile]
    limitation: str | None = None


class ReadinessRequest(BaseModel):
    import_id: UUID
    use_case: GuidedUseCase
    target: str | None = None
    features: list[str] = Field(min_length=1, max_length=128)
    profile: TrainingProfile


class ReadinessResponse(BaseModel):
    ready: bool
    dataset_size: int
    valid_rows: int
    selected_features: list[str]
    target_available: bool
    missingness_rate: float
    time_coverage: str | None
    blocking_issues: list[str]
    profile_summary: str
    deployment_consequence: str


class DemoScenarioName(StrEnum):
    NORMAL = "normal_operation"
    DEGRADATION = "gradual_degradation"
    WARNING = "warning_threshold"
    CRITICAL = "critical_condition"
    SENSOR_FAULT = "sensor_fault"
    DATA_DROPOUT = "data_dropout"
    MAINTENANCE = "maintenance_intervention"
    RECOVERY = "recovery_to_normal"


class DemoScenarioStart(BaseModel):
    factory_id: UUID
    machine_id: UUID
    scenario: DemoScenarioName
    speed: int = Field(default=1)
    idempotency_key: SafeKey

    @model_validator(mode="after")
    def validate_speed(self) -> DemoScenarioStart:
        if self.speed not in {1, 2, 5}:
            raise ValueError("Speed must be 1x, 2x, or 5x.")
        return self


class DemoScenarioResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    factory_id: UUID
    machine_id: UUID
    scenario: str
    status: str
    speed: int
    current_step: int
    generated_readings: int
    state_snapshot: dict[str, object]
    started_at: datetime
    updated_at: datetime
    stopped_at: datetime | None


class LayoutNode(BaseModel):
    machine_id: UUID
    x: float = Field(ge=0, le=100)
    y: float = Field(ge=0, le=100)
    group: str | None = Field(default=None, max_length=64)


class FactoryLayoutUpdate(BaseModel):
    nodes: list[LayoutNode] = Field(max_length=200)


class FactoryLayoutResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    factory_id: UUID
    nodes: list[dict[str, object]]
    version: int
    updated_at: datetime


class ReportType(StrEnum):
    EXECUTIVE_FACTORY = "executive_factory_summary"
    MACHINE_HEALTH = "machine_health"
    ALERT_ACTION = "alert_and_action"
    SHIFT_HANDOVER = "shift_handover"
    DATA_QUALITY = "data_quality"
    MODEL_GOVERNANCE = "model_and_prediction_governance"


class ReportFormat(StrEnum):
    PDF = "pdf"
    XLSX = "xlsx"
    CSV = "csv"


class ReportCreate(BaseModel):
    report_type: ReportType
    format: ReportFormat
    factory_id: UUID | None = None
    period_start: datetime
    period_end: datetime
    idempotency_key: SafeKey


class ReportResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    factory_id: UUID | None
    report_type: str
    format: str
    status: str
    period_start: datetime
    period_end: datetime
    size_bytes: int | None
    summary: dict[str, object]
    safe_error: str | None
    expires_at: datetime | None
    created_at: datetime
    completed_at: datetime | None


class ReportScheduleCreate(BaseModel):
    report_type: ReportType
    format: ReportFormat
    factory_id: UUID | None = None
    period: str = Field(
        pattern=r"^(current_shift|previous_shift|today|last_7_days|last_30_days)$"
    )
    cadence: str = Field(pattern=r"^(daily|weekly|monthly)$")
    timezone: str = Field(default="UTC", min_length=1, max_length=64)
    recipients: list[EmailStr] = Field(min_length=1, max_length=10)
    enabled: bool = False
    idempotency_key: SafeKey


class ReportScheduleResponse(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: UUID
    factory_id: UUID | None
    report_type: str
    format: str
    period: str
    cadence: str
    timezone: str
    recipients: list[str]
    enabled: bool
    last_run_at: datetime | None
    next_run_at: datetime | None
    last_result: str | None
    last_error: str | None


class IncidentSummaryResponse(BaseModel):
    machine_id: UUID
    what_happened: str
    first_detected: datetime | None
    actions_taken: list[str]
    resolution: str
    current_status: str
    follow_up_recommendations: list[str]

File: frontend/src/api/demoExperience.ts
import { apiDownload, apiRequest } from "./client";

export interface DataImport {
  readonly completed_at: string | null;
  readonly created_at: string;
  readonly delimiter: string;
  readonly error_samples: readonly Record<string, unknown>[];
  readonly filename: string;
  readonly id: string;
  readonly imported_rows: number;
  readonly mapping: Record<string, unknown>;
  readonly preview: readonly Record<string, unknown>[];
  readonly progress_percent: number;
  readonly quality_report: Record<string, unknown>;
  readonly rejected_rows: number;
  readonly size_bytes: number;
  readonly status: string;
  readonly total_rows: number;
}

export interface DataImportPage {
  readonly items: readonly DataImport[];
  readonly limit: number;
  readonly offset: number;
  readonly total: number;
}

export interface ImportMapping {
  readonly batch_or_shift?: string;
  readonly features?: readonly string[];
  readonly machine: string;
  readonly operating_state?: string;
  readonly sensor?: string;
  readonly shape: "long" | "wide";
  readonly target?: string;
  readonly timestamp: string;
  readonly value?: string;
}

export function uploadGuidedCsv(
  file: File,
  idempotencyKey: string,
): Promise<DataImport> {
  const body = new FormData();
  body.append("file", file, file.name);
  return apiRequest("/data-onboarding/imports", {
    body,
    headers: { "Idempotency-Key": idempotencyKey },
    method: "POST",
  });
}

export function listGuidedImports(signal?: AbortSignal): Promise<DataImportPage> {
  return apiRequest("/data-onboarding/imports?limit=25", { signal });
}

export function saveImportMapping(
  id: string,
  mapping: ImportMapping,
  delimiter = ",",
): Promise<DataImport> {
  return apiRequest(`/data-onboarding/imports/${id}/mapping`, {
    body: JSON.stringify({ delimiter, has_header: true, mapping }),
    method: "PUT",
  });
}

export function validateImport(id: string): Promise<DataImport> {
  return apiRequest(`/data-onboarding/imports/${id}/validate`, { method: "POST" });
}

export function confirmImport(
  id: string,
  acceptWarnings: boolean,
): Promise<DataImport> {
  return apiRequest(`/data-onboarding/imports/${id}/confirm`, {
    body: JSON.stringify({ accept_warnings: acceptWarnings }),
    method: "POST",
  });
}

export function downloadImportQuality(
  id: string,
  format: "csv" | "json" = "csv",
): Promise<Blob> {
  return apiDownload(`/data-onboarding/imports/${id}/quality.${format}`);
}

export interface ReadinessResult {
  readonly blocking_issues: readonly string[];
  readonly dataset_size: number;
  readonly deployment_consequence: string;
  readonly missingness_rate: number;
  readonly profile_summary: string;
  readonly ready: boolean;
  readonly selected_features: readonly string[];
  readonly target_available: boolean;
  readonly valid_rows: number;
}

export function checkGuidedReadiness(payload: {
  readonly features: readonly string[];
  readonly import_id: string;
  readonly profile: "balanced" | "fast_demo" | "thorough";
  readonly target?: string;
  readonly use_case:
    | "condition_monitoring"
    | "energy_monitoring"
    | "predictive_maintenance"
    | "quality_prediction";
}): Promise<ReadinessResult> {
  return apiRequest("/data-onboarding/guided-ai/readiness", {
    body: JSON.stringify(payload),
    method: "POST",
  });
}

export interface DemoRun {
  readonly current_step: number;
  readonly factory_id: string;
  readonly generated_readings: number;
  readonly id: string;
  readonly machine_id: string;
  readonly scenario: string;
  readonly speed: number;
  readonly state_snapshot: Record<string, unknown>;
  readonly status: string;
}

export function startDemoRun(payload: {
  readonly factory_id: string;
  readonly idempotency_key: string;
  readonly machine_id: string;
  readonly scenario: string;
  readonly speed: number;
}): Promise<DemoRun> {
  return apiRequest("/demo/runs", {
    body: JSON.stringify(payload),
    method: "POST",
  });
}

export function getDemoRun(id: string, signal?: AbortSignal): Promise<DemoRun> {
  return apiRequest(`/demo/runs/${id}`, { signal });
}

export function advanceDemoRun(id: string): Promise<DemoRun> {
  return apiRequest(`/demo/runs/${id}/advance`, { method: "POST" });
}

export function controlDemoRun(
  id: string,
  command: "pause" | "resume" | "stop",
): Promise<DemoRun> {
  return apiRequest(`/demo/runs/${id}/${command}`, { method: "POST" });
}

export function resetDemoRun(
  id: string,
): Promise<{ readonly deleted_records: number }> {
  return apiRequest(`/demo/runs/${id}/reset`, { method: "DELETE" });
}

export interface FactoryLayout {
  readonly factory_id: string;
  readonly id: string;
  readonly nodes: readonly {
    readonly group?: string;
    readonly machine_id: string;
    readonly x: number;
    readonly y: number;
  }[];
  readonly updated_at: string;
  readonly version: number;
}

export function getFactoryLayout(
  factoryId: string,
  signal?: AbortSignal,
): Promise<FactoryLayout | null> {
  return apiRequest(`/demo/factories/${factoryId}/layout`, { signal });
}

export function saveFactoryLayout(
  factoryId: string,
  nodes: FactoryLayout["nodes"],
): Promise<FactoryLayout> {
  return apiRequest(`/demo/factories/${factoryId}/layout`, {
    body: JSON.stringify({ nodes }),
    method: "PUT",
  });
}

export function getTvSummary(
  factoryId: string,
  signal?: AbortSignal,
): Promise<Readonly<Record<string, unknown>>> {
  return apiRequest(`/demo/factories/${factoryId}/tv`, { signal });
}

export interface FactoryMapState {
  readonly action_count: number;
  readonly alert_count: number;
  readonly assessed_at: string | null;
  readonly machine_id: string;
  readonly machine_name: string;
  readonly state: string;
}

export function getFactoryMapState(
  factoryId: string,
  signal?: AbortSignal,
): Promise<readonly FactoryMapState[]> {
  return apiRequest(`/demo/factories/${factoryId}/map-state`, { signal });
}

export interface ExecutiveDashboard {
  readonly definitions: Readonly<Record<string, string>>;
  readonly drill_downs: Readonly<Record<string, string>>;
  readonly limitations: string;
  readonly metrics: Readonly<Record<string, number | string | null>>;
  readonly source: string;
}

export function getExecutiveDashboard(
  signal?: AbortSignal,
): Promise<ExecutiveDashboard> {
  return apiRequest("/reporting/executive-dashboard", { signal });
}

export interface ReportJob {
  readonly completed_at: string | null;
  readonly created_at: string;
  readonly expires_at: string | null;
  readonly factory_id: string | null;
  readonly format: "csv" | "pdf" | "xlsx";
  readonly id: string;
  readonly period_end: string;
  readonly period_start: string;
  readonly report_type: string;
  readonly safe_error: string | null;
  readonly size_bytes: number | null;
  readonly status: string;
  readonly summary: Readonly<Record<string, unknown>>;
}

export function listReports(signal?: AbortSignal): Promise<readonly ReportJob[]> {
  return apiRequest("/reporting/reports", { signal });
}

export function getReport(id: string, signal?: AbortSignal): Promise<ReportJob> {
  return apiRequest(`/reporting/reports/${id}`, { signal });
}

export function createReport(payload: {
  readonly factory_id?: string;
  readonly format: "csv" | "pdf" | "xlsx";
  readonly idempotency_key: string;
  readonly period_end: string;
  readonly period_start: string;
  readonly report_type: string;
}): Promise<ReportJob> {
  return apiRequest("/reporting/reports", {
    body: JSON.stringify(payload),
    method: "POST",
  });
}

export function downloadReport(id: string): Promise<Blob> {
  return apiDownload(`/reporting/reports/${id}/download`);
}
